### PR TITLE
Jonmv/use classes as keys

### PR DESCRIPTION
**Pull request by jonmv**
@baldersheim @bratseth please review.

### CODE REVIEW

**Review by baldersheim — CHANGES_REQUESTED** [review]
This will leave code from old bundles in this cache forever. That is definitely not what you want.
You want to detect that there is a change and drop the cache. and the old code.

**Comment by jonmv** [comment]
If only the map was a `WeakHashMap` ...

**Comment by jonmv** [comment]
Ok, let's clear entries whenever a new class has the same name, instead.

**Comment by jonmv** [comment]
Or would you rather we drop the entire cache whenever this happens?

**Comment by jonmv** [comment]
That's probably the right thing to do.

File: vespajlib/src/main/java/com/yahoo/collections/MethodCache.java [hunk 1]
```diff
@@ -35,7 +35,7 @@ public Method get(Object object) {
     }
     public Method get(Object object, Consumer<String> onPut) {
         Method m = cache.get(object.getClass().getName());
-        if ( ! m.getDeclaringClass().isAssignableFrom(object.getClass())) {
+        if ( m != null && ! m.getDeclaringClass().isAssignableFrom(object.getClass())) {
```
Comment (jonmv) [inline vespajlib/src/main/java/com/yahoo/collections/MethodCache.java:38]: If these calls are slow, we could also just use a composite value of (class, method), and compare class for identity.

**Comment by jonmv** [comment]
Hang on, test on the way.

**Comment by jonmv** [comment]
Ungh ... the weak-ref points to that pair, which isn't owned by anyone else. Need to use refs to the method and object class, of obviously ...

File: vespajlib/src/test/resources/dummy [hunk 2]
```diff
@@ -0,0 +1 @@
+dummy
```
Comment (baldersheim) [inline vespajlib/src/test/resources/dummy:1]: Missing EOL at EOF

File: vespajlib/src/main/java/com/yahoo/collections/MethodCache.java [hunk 3]
```diff
@@ -33,18 +33,23 @@ public void clear() {
...
-        Method m = cache.get(object.getClass().getName());
-        if (m == null) {
-            m = lookupMethod(object);
-            if (m != null) {
-                if (onPut != null)
-                    onPut.accept(object.getClass().getName());
-                cache.put(object.getClass().getName(), m);
-            }
+        Pair<Class<?>, Method> pair = cache.get(object.getClass().getName());
+        if (pair != null && pair.getFirst() != object.getClass()) {
```
Comment (baldersheim) [inline vespajlib/src/main/java/com/yahoo/collections/MethodCache.java:39]: You need a comment similar to the one above to explain the dropping of the cache.

**Review by baldersheim — CHANGES_REQUESTED** [review]
Functionality should be good now, but at least a comment when dropping the cache.
Perhaps also reason for having pair of class and method as value.

File: vespajlib/src/main/java/com/yahoo/collections/MethodCache.java (hunk 3 above)
Comment (jonmv) [reply vespajlib/src/main/java/com/yahoo/collections/MethodCache.java:39]:
```suggestion
        // When changing bundles, you might end up having cached the old method pointing to the old bundle.
        // That will then lead to a class cast exception when invoking the wrong clone method.
        // Whenever we detect a new class with the same name, we therefore drop the entire cache. 
        // This is also the reason for caching the pair of method and original class—not just the method.
        if (pair != null && pair.getFirst() != object.getClass()) {
```

**Pull request merged**